File: scripts/LoadExtendedData.java
import java.io.*;
import java.sql.*;
import java.util.Properties;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class LoadExtendedData {
    public static void main(String[] args) {
        System.out.println("=== Cargar Datos Extendidos en RDS ===\n");
        
        // Cargar variables de entorno desde .env
        Properties env = loadEnvFile(".env");
        
        String dbHost = env.getProperty("DB_HOST");
        String dbPort = env.getProperty("DB_PORT", "5432");
        String dbName = env.getProperty("DB_NAME");
        String dbUser = env.getProperty("DB_USER");
        String dbPassword = env.getProperty("DB_PASSWORD");
        
        if (dbHost == null || dbName == null || dbUser == null || dbPassword == null) {
            System.err.println("Error: Variables de base de datos no configuradas en .env");
            System.err.println("Se requieren: DB_HOST, DB_NAME, DB_USER, DB_PASSWORD");
            System.exit(1);
        }
        
        String sqlFile = "pensamientoComputacional/src/main/resources/data-extended.sql";
        
        if (!new File(sqlFile).exists()) {
            System.err.println("Error: Archivo SQL no encontrado: " + sqlFile);
            System.exit(1);
        }
        
        System.out.println("Conectando a RDS: " + dbHost + ":" + dbPort + "/" + dbName);
        System.out.println("Usuario: " + dbUser);
        System.out.println("Ejecutando script SQL: " + sqlFile + "\n");
        
        String url = "jdbc:postgresql://" + dbHost + ":" + dbPort + "/" + dbName + "?sslmode=require";
        
        try (Connection conn = DriverManager.getConnection(url, dbUser, dbPassword)) {
            conn.setAutoCommit(true);
            
            String sqlContent = readFile(sqlFile);
            String[] statements = splitStatements(sqlContent);
            
            int successCount = 0;
            int errorCount = 0;
            int skipCount = 0;
            
            for (int i = 0; i < statements.length; i++) {
                String statement = statements[i].trim();
                if (statement.isEmpty()) continue;
                
                try (Statement stmt = conn.createStatement()) {
                    stmt.execute(statement);
                    successCount++;
                    System.out.println("  ✓ Statement " + (i + 1) + "/" + statements.length + " ejecutado");
                } catch (SQLException e) {
                    // Ignorar errores de "ya existe" (NOT EXISTS clauses)
                    String errorMsg = e.getMessage().toLowerCase();
                    if (errorMsg.contains("already exists") || 
                        errorMsg.contains("duplicate") ||
                        errorMsg.contains("violates unique constraint")) {
                        skipCount++;
                        System.out.println("  ⚠ Statement " + (i + 1) + " ya existe (ignorado)");
                    } else {
                        errorCount++;
                        System.err.println("  ✗ Error en statement " + (i + 1) + ": " + e.getMessage());
                        String preview = statement.length() > 100 ? statement.substring(0, 100) + "..." : statement;
                        System.err.println("    Statement: " + preview);
                    }
                }
            }
            
            System.out.println("\n✓ Datos extendidos cargados exitosamente");
            System.out.println("  Exitosos: " + successCount);
            System.out.println("  Omitidos (ya existían): " + skipCount);
            if (errorCount > 0) {
                System.out.println("  Errores: " + errorCount);
            }
            
        } catch (SQLException e) {
            System.err.println("✗ Error al conectar a la base de datos: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        } catch (IOException e) {
            System.err.println("✗ Error al leer el archivo SQL: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
        
        System.out.println("\n=== Proceso completado ===");
    }
    
    private static Properties loadEnvFile(String filename) {
        Properties props = new Properties();
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;
                int eqIndex = line.indexOf('=');
                if (eqIndex > 0) {
                    String key = line.substring(0, eqIndex).trim();
                    String value = line.substring(eqIndex + 1).trim();
                    props.setProperty(key, value);
                }
            }
        } catch (IOException e) {
            System.err.println("Advertencia: No se pudo cargar .env: " + e.getMessage());
        }
        return props;
    }
    
    private static String readFile(String filename) throws IOException {
        StringBuilder content = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                // Ignorar líneas de comentarios
                if (!line.trim().startsWith("--")) {
                    content.append(line).append("\n");
                }
            }
        }
        return content.toString();
    }
    
    private static String[] splitStatements(String sql) {
        // Dividir por punto y coma, pero respetar strings y comentarios
        java.util.List<String> statements = new java.util.ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inString = false;
        char stringChar = 0;
        
        for (int i = 0; i < sql.length(); i++) {
            char c = sql.charAt(i);
            
            if (!inString && (c == '\'' || c == '"')) {
                inString = true;
                stringChar = c;
                current.append(c);
            } else if (inString && c == stringChar) {
                inString = false;
                current.append(c);
            } else if (!inString && c == ';') {
                current.append(c);
                String stmt = current.toString().trim();
                if (!stmt.isEmpty()) {
                    statements.add(stmt);
                }
                current = new StringBuilder();
            } else {
                current.append(c);
            }
        }
        
        // Agregar el último statement si no termina con punto y coma
        String last = current.toString().trim();
        if (!last.isEmpty()) {
            statements.add(last);
        }
        
        return statements.toArray(new String[0]);
    }
}

File: scripts/load-extended-data.py
#!/usr/bin/env python3
"""
Script para cargar datos extendidos en RDS usando Python
Alternativa cuando psql no está disponible
"""

import os
import sys
import re

def load_env():
    """Carga variables de entorno desde .env"""
    env_vars = {}
    if os.path.exists('.env'):
        with open('.env', 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    env_vars[key.strip()] = value.strip()
    return env_vars

def check_psycopg2():
    """Verifica si psycopg2 está disponible"""
    try:
        import psycopg2
        return True, psycopg2
    except ImportError:
        return False, None

def execute_sql_with_psycopg2(env_vars, sql_file):
    """Ejecuta SQL usando psycopg2"""
    try:
        import psycopg2
        from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
        
        db_host = env_vars.get('DB_HOST')
        db_port = env_vars.get('DB_PORT', '5432')
        db_name = env_vars.get('DB_NAME')
        db_user = env_vars.get('DB_USER')
        db_password = env_vars.get('DB_PASSWORD')
        
        if not all([db_host, db_name, db_user, db_password]):
            print("Error: Variables de base de datos no configuradas en .env")
            print("Se requieren: DB_HOST, DB_NAME, DB_USER, DB_PASSWORD")
            sys.exit(1)
        
        print(f"Conectando a RDS: {db_host}:{db_port}/{db_name}")
        print(f"Usuario: {db_user}")
        
        conn = psycopg2.connect(
            host=db_host,
            port=db_port,
            database=db_name,
            user=db_user,
            password=db_password,
            sslmode='require'
        )
        
        conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
        cursor = conn.cursor()
        
        print(f"Ejecutando script SQL: {sql_file}")
        
        with open(sql_file, 'r') as f:
            sql_content = f.read()
        
        # Dividir el contenido en statements individuales
        # Remover comentarios y dividir por punto y coma
        statements = []
        current_statement = ""
        
        for line in sql_content.split('\n'):
            # Ignorar líneas de comentarios
            if line.strip().startswith('--'):
                continue
            
            current_statement += line + '\n'
            
            # Si la línea termina con punto y coma, es el final de un statement
            if line.strip().endswith(';'):
                stmt = current_statement.strip()
                if stmt:
                    statements.append(stmt)
                current_statement = ""
        
        # Ejecutar cada statement
        for i, statement in enumerate(statements, 1):
            if statement.strip():
                try:
                    cursor.execute(statement)
                    print(f"  ✓ Statement {i}/{len(statements)} ejecutado")
                except Exception as e:
                    # Ignorar errores de "ya existe" (NOT EXISTS clauses)
                    if 'already exists' in str(e).lower() or 'duplicate' in str(e).lower():
                        print(f"  ⚠ Statement {i} ya existe (ignorado)")
                    else:
                        print(f"  ✗ Error en statement {i}: {e}")
                        print(f"    Statement: {statement[:100]}...")
        
        cursor.close()
        conn.close()
        
        print("\n✓ Datos extendidos cargados exitosamente")
        return True
        
    except Exception as e:
        print(f"✗ Error al conectar o ejecutar SQL: {e}")
        return False

def main():
    print("=== Cargar Datos Extendidos en RDS ===\n")
    
    # Cargar variables de entorno
    env_vars = load_env()
    if not env_vars:
        print("Error: No se pudieron cargar variables de entorno desde .env")
        sys.exit(1)
    
    # Verificar archivo SQL
    sql_file = "pensamientoComputacional/src/main/resources/data-extended.sql"
    if not os.path.exists(sql_file):
        print(f"Error: Archivo SQL no encontrado: {sql_file}")
        sys.exit(1)
    
    # Intentar usar psycopg2
    has_psycopg2, psycopg2_module = check_psycopg2()
    
    if has_psycopg2:
        print("Usando psycopg2 para conectar a PostgreSQL...\n")
        success = execute_sql_with_psycopg2(env_vars, sql_file)
        if success:
            print("\n=== Proceso completado ===")
            sys.exit(0)
        else:
            sys.exit(1)
    else:
        print("psycopg2 no está instalado.")
        print("\nPara instalar psycopg2, ejecuta:")
        print("  pip install psycopg2-binary")
        print("\nO usa el script Java alternativo:")
        print("  ./scripts/load-extended-data-java.sh")
        sys.exit(1)

if __name__ == "__main__":
    main()

